src/mmds/src/token_headers.rs

```diff
@@ -0,0 +1,130 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use micro_http::{HttpHeaderError, RequestError};
+use std::collections::HashMap;
+use std::result::Result;
+
+/// Wrapper over the list of token headers associated with a Request.
+#[derive(Debug, PartialEq)]
+pub struct TokenHeaders {
+    /// The `X-metadata-token` header might be used by HTTP clients to specify a token in order
+    /// to authenticate to the session. This is used for guest requests to MMDS only.
+    x_metadata_token: Option<String>,
+    /// The `X-metadata-token-ttl-seconds` header might be used by HTTP clients to specify
+    /// the expiry time of a token. This is used for PUT requests issued by the guest to MMDS only.
+    x_metadata_token_ttl_seconds: Option<u32>,
+}
+
+impl Default for TokenHeaders {
+    /// Token headers are not present in the request by default.
+    fn default() -> Self {
+        Self {
+            x_metadata_token: None,
+            x_metadata_token_ttl_seconds: None,
+        }
+    }
+}
+
+impl TokenHeaders {
+    /// `X-metadata-token` header.
+    const X_METADATA_TOKEN: &'static str = "X-metadata-token";
+    /// `X-metadata-token-ttl-seconds` header.
+    const X_METADATA_TOKEN_TTL_SECONDS: &'static str = "X-metadata-token-ttl-seconds";
+
+    /// Return `TokenHeaders` from headers map.
+    pub fn try_from(map: &HashMap<String, String>) -> Result<TokenHeaders, RequestError> {
+        let mut headers = Self::default();
+
+        if let Some(token) = map.get(TokenHeaders::X_METADATA_TOKEN) {
+            headers.x_metadata_token = Some(token.to_string());
+        }
+
+        if let Some(value) = map.get(TokenHeaders::X_METADATA_TOKEN_TTL_SECONDS) {
+            match value.parse::<u32>() {
+                Ok(seconds) => {
+                    headers.x_metadata_token_ttl_seconds = Some(seconds);
+                }
+                Err(_) => {
+                    return Err(RequestError::HeaderError(HttpHeaderError::InvalidValue(
+                        TokenHeaders::X_METADATA_TOKEN_TTL_SECONDS.to_string(),
+                        value.to_string(),
+                    )));
+                }
+            }
+        }
```

Comment:
slightly more rustacean, your call if you think it looks better:
```suggestion
        map.get(TokenHeaders::X_METADATA_TOKEN_TTL_SECONDS)
            .and_then(|value| match value.parse::<u32>() {
                Ok(seconds) => Some(Ok(headers.x_metadata_token_ttl_seconds = Some(seconds))),
                Err(_) => Some(Err(RequestError::HeaderError(
                    HttpHeaderError::InvalidValue(
                        TokenHeaders::X_METADATA_TOKEN_TTL_SECONDS.to_string(),
                        value.to_string(),
                    ),
                ))),
            })
            .unwrap()?;
```

Reply:
I agree that your proposal it's more rustacean, but I think it's harder to read. My personal preference would be to leave this as is.

Reply:
personally I prefer the functional approach, but it's only a nit. looks good either way 👍🏻